### PR TITLE
CI: Enhancements & Refactoring

### PR DESCRIPTION
**Pull request by microdev1**
This contains the following:

- split boards dynamically
- refactor workflow names
- conditionally build all languages

### CODE REVIEW

File: tools/build_memory_info.py [hunk 1]
```diff
@@ -51,6 +53,11 @@
 free_flash = firmware_region - used_flash
 used_ram = data + bss
 free_ram = ram_region - used_ram
+
+with open(f"{sys.argv[2]}/firmware.size.json", "w") as f:
+    json.dump({"used_flash": used_flash, "firmware_region": firmware_region}, f)
```
Comment (tannewt) [inline tools/build_memory_info.py:58]: I'm a little worried they aren't detailed enough for complex platforms. Ideally, I'd like to get rid of this python script in favor of correcting linker scripts to use `AT>`. How does this work with imx?

File: tools/ci_set_matrix.py [hunk 2]
```diff
@@ -224,18 +212,15 @@ def get_settings(board):
...
-            for board in last_failed_jobs[job]:
-                boards_to_build.add(board)
+    last_failed_boards = last_failed_jobs.get("boards")
+    if last_failed_boards:
+        boards_to_build.update(last_failed_boards)
 
-    build_boards = bool(boards_to_build)
-    print("Building boards:", build_boards)
-    set_output("build-boards", build_boards)
+    print("Building boards:", bool(boards_to_build))
```
Comment (tannewt) [inline tools/ci_set_matrix.py:220]: Why bool?

File: .github/workflows/re-build-boards.yml [hunk 3]
```diff
@@ -1,11 +1,8 @@
-name: Build boards
+name: Reusable | Build boards
```
Comment (tannewt) [inline .github/workflows/re-build-boards.yml:1]: Are there non-reusable versions of this? I'm not sure why you want the distinction here.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -286,52 +281,15 @@ jobs:
...
-    if: ${{ needs.scheduler.outputs.boards-esp != '[]' }}
-    uses: ./.github/workflows/build-boards.yml
-    secrets: inherit
-    with:
-      platform: esp
-      boards: ${{ needs.scheduler.outputs.boards-esp }}
-      cp-version: ${{ needs.scheduler.outputs.cp-version }}
-
-  riscv:
+  boards:
```
Comment (tannewt) [inline .github/workflows/build.yml:284]:
```suggestion
  ports:
```

File: .github/workflows/re-build-boards.yml [hunk 5]
```diff
@@ -38,6 +35,7 @@ jobs:
       with:
```
Comment (tannewt) [inline .github/workflows/re-build-boards.yml:35]: Rename this job to `board`. That should make the task name change from `Build CI / boards (atmel-samd) / build (8086_commander) (pull_request) ` to `Build CI / ports (atmel-samd) / board (8086_commander) (pull_request)`

**Review by tannewt — CHANGES_REQUESTED** [review]
Thank you for these refinements. I really like the two levels of builds now. One suggestion I had will improve the job naming.

File: .github/workflows/re-build-boards.yml (hunk 3 above)
Comment (microdev1) [reply .github/workflows/re-build-boards.yml:1]: No, this is just a normal workflow. The `workflow_call` event is what makes this reusable.
The reusable prefix in name is added just to separate it from the rest in GitHub UI.

File: tools/ci_set_matrix.py (hunk 2 above)
Comment (microdev1) [reply tools/ci_set_matrix.py:220]: This just logs if any boards are being built.

File: tools/build_memory_info.py (hunk 1 above)
Comment (microdev1) [reply tools/build_memory_info.py:58]: Yes, the size reported by this script and that by the linker differ, on espressif I noticed `script - linker ~ 1500 bytes`.
This doesn't work for imx, all languages are built.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for the update! I like the job names now.

I'd prefer to leave the file names and workflow names as-is and not add the reusable/re prefix.

**Review by tannewt — APPROVED** [review]
Looks great! Thank you!

**Pull request merged**